### PR TITLE
Patron Grant Application

### PR DESCRIPTION
**Pull request by IvanLeshchenko727**
#### Project Abstract

Patron is your one stop smart contract manager for Polkadot ecosystem.

Nowadays, smart contracts are one of the main instruments of development in the blockchain world. As blockchain itself should be open and reliable, we want to create a registry of all available contracts, so everyone could understand the risks and check whether a specific contract is safe or not.

The other side of it - manager of smart contracts, so every developer could manage his own contracts just like using GitHub. It is very important for users to have connection with what developers do, so that will increase trust between them.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application, @IvanLeshchenko727, and welcome back. :)

- The application is currently missing the tech stack(s). 
- For UI grants/deliverables, we usually expect some preliminary design work, such as wireframes or mockups. 
- You didn't provide a `Future Work` section. Can you explain what your long-term plan for Patron is? Are the fees supposed to pay for hosting, maintenance and future development?

File: applications/patron.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+| 0c. | Testing guidelines | Core functionality will be covered by a comprehensive unit test suite. |
+| 0d. | Docker | We will prepare Docker images for users to spin up their own nodes more easily and conveniently. |
+| 0e. | Article | We will publish an article that explains the achievements done as part of the grant. |
+| 1a. | Backend storage | Backend implementation with contract discovery and persistent storage. |
+| 1b. | Sync server | A separate server that catches new contract deployments and events will be implemented. |
+| 1c. | Smart contract builder | Immutable, pre-configured smart contract builders are to be implemented for verified smart contract deployment. |
+| 2a. | Web UI | A simple web UI will be implemented to expose Patron functionality. |
+| 2b. | Detailed contract information | Frontend to display detailed contract info (as well as verification status) will be implemented. |
+| 2c. | User authentication | We will implement a web3-oriented authentication solution |
+| 3a. | Developer CLI utility | A deployment workflow unified, developer-oriented CLI utility will be implemented. |
```
Comment (semuelle) [inline applications/patron.md:257]: What does "deployment workflow unified" mean?

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the updates, @IvanLeshchenko727. 
This application covers so many different features that it's difficult to see through it. But could you add a bit more technical details for the CLI (what's in M1, what's in M2?) and verification module?

Generally speaking, I think the verification module and deployment tool has greater appeal to the ecosystem than the contract manager and another explorer. Also, at the moment this would pay 13,856 USD per person-month, which is unusually high.

Either way, I am going the share the application with the rest of the team so they can chime in.

**Comment by keeganquigley** [comment]
Thanks for the application @IvanLeshchenko727 a couple of questions:
- You mentioned that current CLI tools only have basic features; could you give a few examples of your solution's features over existing ones? For example, what features will your CLI tool have compared to [Swanky suite](https://github.com/AstarNetwork/swanky-cli) by Astar?
- I like the idea of reproducible builds, but can you explain what you mean when you say the containers are provided by the user?
- Is the verification module verifying the cryptographic hash of the build? Or can you explain more about what its function is and what verified WASM blob means exactly?
- What will the fee to Patron be paid in?
- What will the sync server be implemented in?

**Comment by keeganquigley** [comment]
Thanks @IvanLeshchenko727 I definitely think this could be useful and generally approve, but I agree the monthly price seems high. Especially for milestone 2, would you be willing to reduce the scope/price?

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. I have a few questions/comments:

- Could you add the default deliverables also to the second milestone?
- I agree that it’s relatively expensive. Did you already try to get this funded via the treasury or https://use.ink/ubator/?

**Comment by freehere107** [comment]
@IvanLeshchenko727  Hey, Since subscan is also been paying attention to wasm recently, I have some questions. Can you describe the technical details of contract verification? In addition, I did not find the verification contract a relevant deliverable.

**Comment by coreggon11** [comment]
Hi @takahser 

Thank you for the questions, we'll get back to them

> Also, in general, I feel like the price is rather high.

Could you elaborate on why you feel the price is high? We use the same rates as in our previous grants.

**Comment by semuelle** [comment]
Sorry for the late reply, @IvanLeshchenko727. I am still in favor of focusing on the verification module and deployment tool. But I'll ping the rest of the committee to see what they say.

**Comment by keeganquigley** [comment]
> Hi! We have re-evaluated the proposal and realized that we forgot to include Product owner work and Designer work. Also in order to optimize the budget and include your recommendations we have reduced the hourly rate and removed scripting and social functionality from M2.
Please see the updated proposal and looking forward towards your feedback.

Thanks for the making the changes. It sounds like the committee is generally in favor but the price may still be too high. Regarding this, it's hard to give an exact number, but given the amount of grants 727 has received in the past, perhaps it would be best to further reduce the scope/price to focus on a PoC for the tools, and then go for treasury funding afterwards, as they may be willing to accept the higher rates.

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I will share it again with everyone.

**Comment by randombishop** [comment]
Hi @IvanLeshchenko727 and thank you for this application!

I only have one question about deliverable `1c. Smart contract builder`: isn't this somewhat similar to what these 2 projects achieve? 
[ink-verifier-server](https://github.com/web3labs/ink-verifier-server)
[ink-verifier-image](https://github.com/web3labs/ink-verifier-image)

Will you re-use those? And if no, can you elaborate on the difference in your approach?

Thanks!

**Comment by randombishop** [comment]
Thank you @IvanLeshchenko727 .
Happy to approve this application.

**Review by randombishop — APPROVED** [review]

**Review by dsm-w3f — APPROVED** [review]
Generally, I'm happy to support the project since code verification is useful and would help to ensure trust in smart contracts. Some small things to notice and that could be considered to improve the proposed system:
- Pay for use. Correct me if I'm wrong, but even EtherScan and Defender are not paid at a certain level of usage. If the intention is to prevent abuse a Stake would be used instead of payment. The freemium model also could be considered for a hosted service. 
- This application seems very likely to become a business after the delivery. It is a requirement that all deliverables should be open source and the user should be capable to run and host this service by themself if they want to. Don't make sense to pay if I want to self-host a site that shows the verification of my contracts.

**Pull request merged**